### PR TITLE
feat: Allow chain specs config dep groups

### PR DESCRIPTION
**Pull request by TheWaWaR**
- Allow chain specs config dep groups in `spec.toml`
- `docs/hashes.toml` with more information (`tx_hash`, `dep_groups`)

### CODE REVIEW

File: ckb-bin/src/subcommand/cli/hashes.rs [hunk 1]
```diff
@@ -90,10 +116,22 @@ pub fn hashes<'m>(root_dir: PathBuf, matches: &ArgMatches<'m>) -> Result<(), Exi
...
         config.chain.spec.absolutize(&root_dir);
         let chain_spec = ChainSpec::load_from(&config.chain.spec).map_err(to_config_error)?;
-        specs.insert(chain_spec.name.clone(), chain_spec.try_into()?);
+        let spec_name = chain_spec.name.clone();
+        let spec_hashes: SpecHashes = chain_spec.try_into()?;
+        specs.push((spec_name, spec_hashes));
     }
 
-    print!("{}", toml::to_string(&specs).unwrap());
+    println!();
```
Comment (u2) [inline ckb-bin/src/subcommand/cli/hashes.rs]: It's intended?

File: docs/hashes.toml [hunk 2]
```diff
@@ -1,72 +1,120 @@
...
 [[ckb_dev.system_cells]]
 path = "Bundled(specs/cells/secp256k1_ripemd160_sha256_sighash_all)"
+tx_hash = "0x57d4f9e30b71e949a126c4b9094cfecdb95b28993f638f74a9aba705c8a33389"
 index = 4
 code_hash = "0x1acf88382192fbf76fc5a457a11cc393a90d84a46c5e0d7ceebfd1d0dec871e4"
 
-[ckb_staging]
-genesis = "0xa0e8f608ea2cc11ce4eb583ed7ac2228b9dfbe71eadd960085f6c27858e2643f"
-cellbase = "0x480637c97e3b43aceb4e4d60b7fad293f60b7938054d9697192524b4e158395a"
+[[ckb_dev.dep_groups]]
```
Comment (doitian) [inline docs/hashes.toml]: I think it is better to add the info which system cells are included in this group.

File: resource/specs/staging.toml [hunk 3]
```diff
@@ -57,6 +57,17 @@ lock.code_hash = "0x54811ce986d5c3e57eaafab22cdd080e32209e39590e204a99b32935f835
 lock.args = ["0x57ccb07be6875f61d93636b0ee11b675494627d2"]
 lock.hash_type = "Data"
 
+# Dep group cells
```
Comment (doitian) [inline resource/specs/staging.toml]: Put it after `[genesis.system_cells.lock]`.

File: spec/src/lib.rs [hunk 4]
```diff
@@ -282,40 +281,33 @@ impl Genesis {
...
-        let outputs = vec![cell_blake160, cell_ripemd160];
-        let outputs_data = vec![data_blake160.pack(), data_ripemd160.pack()];
+        let (outputs, outputs_data): (Vec<_>, Vec<_>) = self
+            .dep_groups
+            .values()
+            .map(|files| {
+                let out_points: Vec<_> = files
+                    .iter()
+                    .map(|res| {
+                        let data: Bytes = res.get()?.into_owned().into();
```
Comment (doitian) [inline spec/src/lib.rs]: I think it is a bit overhead to read the data to just check what's the position of it in the cellbase. We also can search the res in the system cells files list. Yes, it means the user must reference the same resource via the same path, but it is acceptable.

File: ckb-bin/src/subcommand/cli/hashes.rs (hunk 1 above)
Comment (TheWaWaR) [reply ckb-bin/src/subcommand/cli/hashes.rs]: Yes

File: docs/hashes.toml (hunk 2 above)
Comment (TheWaWaR) [reply docs/hashes.toml]: ok

File: resource/specs/staging.toml (hunk 3 above)
Comment (TheWaWaR) [reply resource/specs/staging.toml]: Fine with me. My original thought is dep groups are in 2nd transaction.

File: spec/src/lib.rs (hunk 4 above)
Comment (TheWaWaR) [reply spec/src/lib.rs]: Or, we can cache the hash result?
`HashMap<Resource, H256>`

**Comment by TheWaWaR** [comment]
bors cancel

**Comment by doitian** [comment]
Merged in https://github.com/nervosnetwork/ckb/commit/644fdaa1a268be262576a97d1b0917c437e95c82

**Pull request closed without merging**